### PR TITLE
Erasing of the list of workspaces before adding new ones. During the initialization.

### PR DESCRIPTION
**Pull request by Khomutov**
this will allow to change BaseDirectory property and than call method Initializeto reload the list of workspaces.

### CODE REVIEW

**Comment by GeertvanHorrik** [comment]
Please add unit tests to automate the testing of multiple schemes.

File: src/Orc.WorkspaceManagement/Managers/WorkspaceManager.cs [hunk 1]
```diff
@@ -119,6 +119,8 @@ public IWorkspace Workspace
 
             if (Directory.Exists(baseDirectory))
             {
+                _workspaces.Clear();
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement/Managers/WorkspaceManager.cs]: I think the workspaces should be cleared anyway (even when directory doesn't exist). I think it's important to put the clear outside the if (even before the Initializing method)

**Comment by GeertvanHorrik** [comment]
I think we should also make switching between "workspace schema sets" very easy with an extension methods (need to think of a better method name though):

```
public static async Task SwitchWorkspaceSchemeDirectory(this IWorkspaceManager workspaceManager, string newDirectory)
{
    Argument.IsNotNull(() => workspaceManager);

    workspaceManager.BaseDirectory = newDirectory;
    await workspaceManager.Initialize();
}
```

**Comment by Khomutov** [comment]
I think better not to use `workspaceManager.Initialize();` but use `workspaceManager.Initialize(true, "Default");` Because if new BaseDirectory will be empty so we will get empty list of workspaces. Even without default one.

**Comment by Khomutov** [comment]
I mean need to create something like this:
```
public static async Task SwitchWorkspaceSchemeDirectory(this IWorkspaceManager workspaceManager, string newDirectory, bool createDwfaultIfEmpty = false, string defaultTitle = "Default")
{
    Argument.IsNotNull(() => workspaceManager);

    workspaceManager.BaseDirectory = newDirectory;
    await workspaceManager.Initialize(createDwfaultIfEmpty, defaultTitle);
}
```

File: src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="WorkspaceManagerExtensionsFacts.cs" company="Orchestra development team">
+//   Copyright (c) 2008 - 2015 Orchestra development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Orc.WorkspaceManagement.Test.Extensions
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs]: Extensions is not a namespace

File: src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        public class SetWorkspaceSchemesDirectoryMethod
+        {
+            [TestCase]
+            public void BaseDirectoryChanged()
+            {
+                const string someDirectoryName = "Some directory";
+
+                var workspaceManager = Factories.WorkspaceManager.WithEmptyInitializer();
+
+                workspaceManager.SetWorkspaceSchemesDirectory(someDirectoryName).Wait();
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs]: Why didn't you make the method async Task? Then you can use await.

File: src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            [TestCase("1", "2")]
+            public void LoadsCorrectWorkspaces(params string[] titles)
+            {
+                var mock = new Mock<IWorkspacesStorageService>();
+
+                mock.Setup(x => x.LoadWorkspaces(It.IsAny<string>())).Returns(titles.Select(t => new Workspace {Title = t}));
+
+                var workspaceManager = Factories.WorkspaceManager.WithEmptyInitializer(mock.Object);
+
+                workspaceManager.SetWorkspaceSchemesDirectory("Some directory").Wait();
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs]: async / await

File: src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                var mock = new Mock<IWorkspacesStorageService>();
+
+                const string emptyDirectory = "Empty directory";
+
+                mock.Setup(x => x.LoadWorkspaces(emptyDirectory)).Returns(Enumerable.Empty<IWorkspace>());
+
+                var workspaceManager = Factories.WorkspaceManager.WithEmptyInitializer(mock.Object);
+
+                workspaceManager.SetWorkspaceSchemesDirectory(emptyDirectory).Wait();
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement.Test/Extensions/WorkspaceManagerExtensionsFacts.cs]: async / await

File: src/Orc.WorkspaceManagement.Test/Helpers/Factories.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Factories.cs" company="Orchestra development team">
+//   Copyright (c) 2008 - 2015 Orchestra development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace Orc.WorkspaceManagement.Test.Helpers
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement.Test/Helpers/Factories.cs]: Helpers is not a namespace

**Comment by GeertvanHorrik** [comment]
Excellent work on introducing the factories.

File: src/Orc.WorkspaceManagement/Managers/WorkspaceManager.cs [hunk 7]
```diff
@@ -117,40 +119,18 @@ public IWorkspace Workspace
...
-                    }
-                    catch (Exception ex)
-                    {
-                        Log.Error(ex, "Failed to load workspace from '{0}'", workspaceFile);
-                    }
-                }
-            }
+            _workspaces.Clear();
+
+            _workspaces.AddRange(_workspacesStorageService.LoadWorkspaces(baseDirectory));
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement/Managers/WorkspaceManager.cs]: Try not to put too many calls on the same line. Split them up (easier to debug and understand for next developers)

File: src/Orc.WorkspaceManagement/ModuleInitializer.cs [hunk 8]
```diff
@@ -1,5 +1,6 @@
 ﻿using Catel.IoC;
 using Orc.WorkspaceManagement;
+using Orc.WorkspaceManagement.Services;
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement/ModuleInitializer.cs]: Please remove services namespace

File: src/Orc.WorkspaceManagement/Services/Interfaces/IWorkspacesStorageService.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Orc.WorkspaceManagement.Services
```
Comment (GeertvanHorrik) [inline src/Orc.WorkspaceManagement/Services/Interfaces/IWorkspacesStorageService.cs]: Remove .Services (we use a single namespace (Orc.WorkspaceManagement) for very simple packages).

**Pull request merged**